test/cpp/api/dataloader.cpp

```diff
@@ -568,7 +607,7 @@ struct TestIndexDataset
 
 struct TestIndexSampler : public samplers::Sampler<TestIndex> {
   explicit TestIndexSampler(size_t size) : size_(size) {}
-  void reset() override {}
+  void reset(optional<size_t> new_size) override {}
```

Comment:
This is bit more complicated with templates. Having defaults at the Sampler does not mean the derived classes will have them when someone use it with the derived class. I did change the three samplers to take nullopt as default so that we don't have to change the tests.